### PR TITLE
feat(integrations): add litellm integration

### PR DESCRIPTION
**Pull request by constantinius**
Add a first implementation of the litellm integration, supporting completion and embeddings

Closes https://linear.app/getsentry/issue/PY-1828/add-agent-monitoring-support-for-litellm

### CODE REVIEW

File: sentry_sdk/integrations/litellm.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        if _input_callback not in litellm.input_callback:
+            litellm.input_callback.append(_input_callback)
+
+        litellm.success_callback = litellm.success_callback or []
+        if _success_callback not in litellm.success_callback:
+            litellm.success_callback.append(_success_callback)
+
+        litellm.failure_callback = litellm.failure_callback or []
+        if _failure_callback not in litellm.failure_callback:
+            litellm.failure_callback.append(_failure_callback)
```
Comment (constantinius) [inline sentry_sdk/integrations/litellm.py:251]: It seems as if both `success_callback` and `failure_callback` are run in a thread, which might finish after `completion` returns. As the span is closed in either callback, it may occur that the span is finished after the surrounding transaction does, resulting it being absent completely. This should definitely be pointed out somewhere.

File: sentry_sdk/integrations/litellm.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    # Record other parameters
+    params = {
+        "model": SPANDATA.GEN_AI_REQUEST_MODEL,
+        "stream": SPANDATA.GEN_AI_RESPONSE_STREAMING,
+        "max_tokens": SPANDATA.GEN_AI_REQUEST_MAX_TOKENS,
+        "presence_penalty": SPANDATA.GEN_AI_REQUEST_PRESENCE_PENALTY,
+        "frequency_penalty": SPANDATA.GEN_AI_REQUEST_FREQUENCY_PENALTY,
+        "temperature": SPANDATA.GEN_AI_REQUEST_TEMPERATURE,
+        "top_p": SPANDATA.GEN_AI_REQUEST_TOP_P,
+    }
```
Comment (constantinius) [inline sentry_sdk/integrations/litellm.py:88]: It is not clear where to actually put these parameters in the arguments to `completion`

File: sentry_sdk/integrations/litellm.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        name=f"{operation} {model}",
+        origin=LiteLLMIntegration.origin,
+    )
+    span.__enter__()
+
+    # Store span for later
+    kwargs["_sentry_span"] = span
+
+    # Set basic data
+    set_data_normalized(span, SPANDATA.GEN_AI_SYSTEM, "litellm")
```
Comment (constantinius) [inline sentry_sdk/integrations/litellm.py]: Should this be `litellm`? Or the actual provider being used? (anthropic, openai, ...)

File: sentry_sdk/integrations/litellm.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    span = get_start_span_function()(
+        op=(
+            consts.OP.GEN_AI_CHAT
+            if operation == "chat"
+            else consts.OP.GEN_AI_EMBEDDINGS
+        ),
+        name=f"{operation} {model}",
+        origin=LiteLLMIntegration.origin,
+    )
+    span.__enter__()
```
Comment (constantinius) [inline sentry_sdk/integrations/litellm.py:64]: We start a transaction if we don't have one ready yet.

File: sentry_sdk/integrations/__init__.py [hunk 5]
```diff
@@ -97,6 +97,7 @@ def iter_default_integrations(with_auto_enabling_integrations):
     "sentry_sdk.integrations.langchain.LangchainIntegration",
     "sentry_sdk.integrations.langgraph.LanggraphIntegration",
     "sentry_sdk.integrations.litestar.LitestarIntegration",
+    "sentry_sdk.integrations.litellm.LiteLLMIntegration",
```
Comment (sentrivana) [inline sentry_sdk/integrations/__init__.py]: I'd prefer we don't enable the integration by default from the start for several reasons:

- It's good practice for ironing out issues before everyone gets the integration OOTB
- It looks like LiteLLM itself has some form of Sentry integration https://github.com/BerriAI/litellm/blob/0d738b28998dc5bba42653083b32e5f333b1ff5a/docs/my-website/docs/proxy/logging.md?plain=1#L2414 so we might end up double instrumenting
- To get some feedback from folks whether they're experiencing any timing issues re: callbacks in threads

File: sentry_sdk/integrations/litellm.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        return "azure"
+    elif "bedrock/" in model:
+        return "bedrock"
+    elif "ollama/" in model:
+        return "ollama"
+    else:
+        # Try to use LiteLLM's internal provider detection if available
+        try:
+            if hasattr(litellm, "get_llm_provider"):
+                provider_info = litellm.get_llm_provider(model)
```
Comment (sentrivana) [inline sentry_sdk/integrations/litellm.py]: I was trying to find out if calling this function could potentially have unintended side-effects (like making a network call I suppose, since it might, for some reason, return a `BadRequestError`), and I quickly got overwhelmed by the [spaghettiness of the code](https://github.com/BerriAI/litellm/blob/0d738b28998dc5bba42653083b32e5f333b1ff5a/litellm/litellm_core_utils/get_llm_provider_logic.py#L88). It's almost impossible to parse.

Hopefully using this is ok side-effect wise, since we only provide it with a model (and not api key/base etc.).

File: sentry_sdk/integrations/litellm.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+        # Capture the exception
+        event, hint = event_from_exception(
+            exception,
+            client_options=sentry_sdk.get_client().options,
+            mechanism={"type": "litellm", "handled": False},
+        )
+        sentry_sdk.capture_event(event, hint=hint)
+    finally:
+        # Always finish the span and clean up
+        span.__exit__(None, None, None)
```
Comment (sentrivana) [inline sentry_sdk/integrations/litellm.py]: The `__exit__` here should be given the exception so that we set the span status correctly to failed.

File: sentry_sdk/integrations/litellm.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+
+    Usage:
+
+    ```python
+    import litellm
+    import sentry_sdk
+
+    # Initialize Sentry with the LiteLLM integration
+    sentry_sdk.init(
+        dsn="your-dsn",
```
Comment (sentrivana) [inline sentry_sdk/integrations/litellm.py:209]: I'd also include `send_default_pii=True` in the example here so that folks get the most out of the box by default. Maybe with a small comment what it actually controls in this integration, like you did for `include_prompts`.

File: sentry_sdk/integrations/litellm.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+            if operation == "chat"
+            else consts.OP.GEN_AI_EMBEDDINGS
+        ),
+        name=f"{operation} {model}",
+        origin=LiteLLMIntegration.origin,
+    )
+    span.__enter__()
+
+    # Store span for later
+    kwargs["_sentry_span"] = span
```
Comment (sentrivana) [inline sentry_sdk/integrations/litellm.py]: I'm wondering if we can store this in `litellm_params.metadata` instead? At least [on this page](https://docs.litellm.ai/docs/observability/custom_callback#whats-available-in-kwargs) it says that is supposed to be "your custom metadata", which could be less intrusive than adding a kwarg.

File: sentry_sdk/integrations/litellm.py (hunk 3 above)
Comment (sentrivana) [reply sentry_sdk/integrations/litellm.py]: The actual provider as far as I understand. OTel says: 

> The Generative AI provider as identified by the client or server instrumentation.

The attribute itself is [deprecated](https://opentelemetry.io/docs/specs/semconv/registry/attributes/gen-ai/#gen-ai-system) and was renamed to [`gen_ai.provider.name`](https://opentelemetry.io/docs/specs/semconv/registry/attributes/gen-ai/#gen-ai-provider-name). I'm not sure what the plan regarding keeping up with OTel's semantic conventions is; you probably know better if we want to use the old or the new way.

File: sentry_sdk/integrations/litellm.py (hunk 2 above)
Comment (sentrivana) [reply sentry_sdk/integrations/litellm.py:88]: Don't understand this comment, can you elaborate? What do the params have to do with `completion`?

File: sentry_sdk/integrations/litellm.py (hunk 1 above)
Comment (sentrivana) [reply sentry_sdk/integrations/litellm.py:251]: There is definitely the potential for a timing issue but I don't see a way around it at the moment since the LiteLLM integration might not be in control of the overarching transaction.

From your testing when developing this, was this a real issue when something like a web framework was managing the transaction?

File: sentry_sdk/integrations/litellm.py (hunk 9 above)
Comment (constantinius) [reply sentry_sdk/integrations/litellm.py]: Thanks, I missed that!

File: sentry_sdk/integrations/litellm.py (hunk 6 above)
Comment (constantinius) [reply sentry_sdk/integrations/litellm.py]: It does raise these Exceptions but manually and not from an API client. The function seems to be only running locally, so it should be fine. In fact I removed the whole of `_get_provider_from_model` in favour of the values returned by `litellm.get_llm_provider`

File: sentry_sdk/integrations/litellm.py (hunk 3 above)
Comment (constantinius) [reply sentry_sdk/integrations/litellm.py]: Will use the actual provider.
We do have a summary issue for upgrading to [OTel 1.37](https://github.com/getsentry/sentry-python/issues/4753), so I'll keep the old one for now.

File: sentry_sdk/integrations/litellm.py (hunk 2 above)
Comment (constantinius) [reply sentry_sdk/integrations/litellm.py:88]: `completion` takes quite generic `kwargs` that are then passed on to the model provider API. These above are used for `openai` (at least I suspect that this is where and how we retrieve them)

File: sentry_sdk/integrations/litellm.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+    # Record LiteLLM-specific parameters
+    litellm_params = {
+        "api_base": kwargs.get("api_base"),
+        "api_version": kwargs.get("api_version"),
+        "custom_llm_provider": kwargs.get("custom_llm_provider"),
+    }
+    for key, value in litellm_params.items():
+        if value is not None:
+            set_data_normalized(span, f"gen_ai.litellm.{key}", value)
+
```
[the remark this answers is not in the document]
Comment (constantinius) [reply sentry_sdk/integrations/litellm.py]: No, the callback is wrapped in `try`/`except` in litellm itself

File: sentry_sdk/integrations/litellm.py (hunk 1 above)
Comment (constantinius) [reply sentry_sdk/integrations/litellm.py:251]: It was only an issue when writing code like this:

```python
with sentry_sdk.start_transaction(...):
    result = completion(...)
```

When using it in a framework (tried with FastAPI) I could not reproduce this error.

**Comment by sentrivana** [comment]
TY for incorporating my feedback, LGTM. From my POV there are 2 outstanding points:
- Adding a test suite & the minimum supported version in `integrations/__init__.py`: see https://github.com/getsentry/sentry-python/blob/master/scripts/populate_tox/README.md#add-a-new-test-suite for more info
- If we want the SDK to have an `litellm` extra as mentioned in [the docs PR](https://github.com/getsentry/sentry-docs/pull/15091/files), it needs to be added to `setup.py`

LMK if you need support with any of these or if you want my feedback on anything else.

**Comment by constantinius** [comment]
> * Adding a test suite & the minimum supported version in `integrations/__init__.py`: see [`master`/scripts/populate_tox/README.md#add-a-new-test-suite](https://github.com/getsentry/sentry-python/blob/master/scripts/populate_tox/README.md?rgh-link-date=2025-10-02T08%3A17%3A24Z#add-a-new-test-suite) for more info

I've added tests and added the minimum versions to the list. I hope that works out.

> * If we want the SDK to have an `litellm` extra as mentioned in [the docs PR](https://github.com/getsentry/sentry-docs/pull/15091/files), it needs to be added to `setup.py`

Done

**Review by sentrivana — APPROVED** [review]
Took the liberty to add the missing test setup (the tests were not running in ci) and fix mypy's complaints, so if ci is green we should be good to go.
